packages/vite-plugin-cloudflare/src/plugins/virtual-modules.ts

```diff
@@ -31,11 +35,56 @@ export const virtualModulesPlugin = createPlugin("virtual-modules", (ctx) => {
 
 				return `
 ${nodeJsCompat ? nodeJsCompat.injectGlobalCode() : ""}
+import { getExportTypes } from "${VIRTUAL_EXPORT_TYPES}";
 import * as mod from "${VIRTUAL_USER_ENTRY}";
 export * from "${VIRTUAL_USER_ENTRY}";
 export default mod.default ?? {};
 if (import.meta.hot) {
-	import.meta.hot.accept();
+	import.meta.hot.accept((module) => {
+		const exportTypes = getExportTypes(module);
+		import.meta.hot.send("vite-plugin-cloudflare:worker-export-types", exportTypes);
+	});
+}
+				`;
+			}
+
+			if (id === `\0${VIRTUAL_EXPORT_TYPES}`) {
+				return `
+import {
+	WorkerEntrypoint,
+	DurableObject,
+	WorkflowEntrypoint,
+} from "cloudflare:workers";
+
+const baseClasses = new Map([
+	["WorkerEntrypoint", WorkerEntrypoint],
+	["DurableObject", DurableObject],
+	["WorkflowEntrypoint", WorkflowEntrypoint],
+]);
+
+export function getExportTypes(module) {
+	const exportTypes = {};
+
+	for (const [key, value] of Object.entries(module)) {
+		if (key === "default") {
+			continue;
+		}
+
+		let exportType
+
+		if (typeof value === "function") {
```

Comment:
I think it would be good to warn the user if they have other export types in their entry module. That could be done in a follow-up PR.